### PR TITLE
Deprecate torch.lu

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73878
* #73877
* #74046
* #73806
* __->__ #73804
* #73803
* #72935

**BC-breaking note**:

This PR deprecates `torch.lu` in favor of `torch.linalg.lu_factor`.
A upgrade guide is added to the documentation for `torch.lu`.

Note this PR DOES NOT remove `torch.lu`.

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -1795,6 +1795,17 @@ std::tuple<Tensor, Tensor> linalg_lu_factor(const Tensor& A, bool pivot) {
 
 // TODO Deprecate this function in favour of linalg_lu_factor_ex
 std::tuple<Tensor, Tensor, Tensor> _lu_with_info(const Tensor& self, bool compute_pivots, bool) {
+   TORCH_WARN_ONCE(
+    "torch.lu is deprecated in favor of torch.linalg.lu_factor / torch.linalg.lu_factor_ex and will be ",
+    "removed in a future PyTorch release.\n",
+    "LU, pivots, info = torch.lu(A, compute_pivots)\n",
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1801]:
```suggestion
    "LU, pivots = torch.lu(A, compute_pivots)\n",
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -1795,6 +1795,17 @@ std::tuple<Tensor, Tensor> linalg_lu_factor(const Tensor& A, bool pivot) {
...
+    "should be replaced with\n",
+    "LU, pivots = torch.linalg.lu_factor(A, compute_pivots)\n",
+    "and\n",
+    "LU, pivots, info = torch.lu(A, compute_pivots, get_infos=True)\n",
+    "should be replaced with\n",
+    "LU, pivots, info = torch.linalg.lu_factor_ex(A, compute_pivots, false)"
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1807]: `check_errors` is False by default.
```suggestion
    "LU, pivots, info = torch.linalg.lu_factor_ex(A, compute_pivots)"
```

File: torch/functional.py [hunk 3]
```diff
@@ -1673,6 +1673,21 @@ def _lu_impl(A, pivot=True, get_infos=False, out=None):
     pivots of :attr:`A`.  Pivoting is done if :attr:`pivot` is set to
     ``True``.
 
+    .. warning::
+
+        :func:`torch.lu` is deprecated in favor of :func:`torch.linalg.lu_factor`
+        and :func:`torch.linalg.lu_factor_ex`. :func:`torch.lu` will be removed in a
+        future PyTorch release.
+        ``LU, pivots, info = torch.lu(A, compute_pivots)`` should be replaced with
```
Comment (IvanYashchuk) [inline torch/functional.py:1681]:
```suggestion
        ``LU, pivots = torch.lu(A, compute_pivots)`` should be replaced with
```

File: torch/functional.py [hunk 4]
```diff
@@ -1673,6 +1673,21 @@ def _lu_impl(A, pivot=True, get_infos=False, out=None):
...
+
+        .. code:: python
+            LU, pivots = torch.linalg.lu_factor(A, compute_pivots)
+
+        ``LU, pivots, info = torch.lu(A, compute_pivots, get_infos=True)`` should be replaced with
+
+        .. code:: python
+            LU, pivots, info = torch.linalg.lu_factor_ex(A, compute_pivots, false)
```
Comment (IvanYashchuk) [inline torch/functional.py:1689]:
```suggestion
            LU, pivots, info = torch.linalg.lu_factor_ex(A, compute_pivots)
```

**Comment by IvanYashchuk** [comment]
There are a few problems with the suggested code in the warning message.

**Comment by Lezcano** [comment]
Fixed, thanks for spotting these!

**Review by IvanYashchuk — APPROVED** [review]

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -2256,6 +2256,17 @@ std::tuple<Tensor, Tensor> linalg_lu_factor(const Tensor& A, bool pivot) {
 
 // TODO Deprecate this function in favour of linalg_lu_factor_ex
 std::tuple<Tensor, Tensor, Tensor> _lu_with_info(const Tensor& self, bool compute_pivots, bool) {
+   TORCH_WARN_ONCE(
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:2259]: Nice warn once use

**Review by mruberry — APPROVED** [review]
Nice deprecation - stamped!

**Comment by Lezcano** [comment]
Relanding in https://github.com/pytorch/pytorch/pull/77636

**Pull request closed without merging**